### PR TITLE
feat(v2): proper realtime hot reload with react-hot-loader

### PR DESCRIPTION
**Pull request by endiliey**
## TL;DR See Test Plan

@yangshun mentioned to me on Discord that the hot-reload is actually not working. So I did some investigation and it was indeed not working. It fallbacks to page refresh.

It was because I forgot to correctly configure the `module.hot` 😢.  See https://webpack.js.org/guides/hot-module-replacement/.

Anyway. I learned a lot about Hot Module Replacement. Things are slightly more complicated on our side because ***in development: we are a SPA*** and ***in production: we are SSR + SPA*.** There is a reason why tools like `create-react-app` is very popular, hahaha. They settle this kind of things for us

However, I remembered [Dan Abramov - Live React: Hot Reloading with Time Travel at react-europe 2015](https://www.youtube.com/watch?v=xsSnOQynTHs), so in this PR I added https://github.com/gaearon/react-hot-loader to take the hot reload into next level. (Real-time hot reload)

Other changes:
- Add `--hot-only` CLI options to let user disable fallback to full page refresh if hot-reload is not working
- Use express.static to follow v1 way, contentBase should be the default publicPath
- Add some TODO plugin hooks on devserver

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/Docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

### Development:

**Before**

Notice that there is a flash of screen. This is in fact a full page refresh. This is quite similar to v1 way of live reload. 
not hot reload

**After**

Notice that there is no flash of screen. This is a hot reload. In fact, it's realtime
hot reload fix

### Production:
Netlify - https://deploy-preview-1336--docusaurus-2.netlify.com/
 
## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/facebook/docusaurus, and link to your PR here.)

### CODE REVIEW

File: packages/docusaurus/lib/commands/start.js [hunk 1]
```diff
@@ -119,7 +123,17 @@ module.exports = async function start(siteDir, cliOptions = {}) {
     disableHostCheck: true,
     overlay: false,
     host,
-    contentBase: path.resolve(siteDir, 'static'),
+    // https://webpack.js.org/configuration/dev-server/#devserverbefore
+    // eslint-disable-next-line
```
Comment (endiliey) [inline packages/docusaurus/lib/commands/start.js]: Because it complains `server` not being used :( eslint-disable it for now

File: packages/docusaurus/lib/commands/start.js [hunk 2]
```diff
@@ -119,7 +123,17 @@ module.exports = async function start(siteDir, cliOptions = {}) {
...
+    before(app, server) {
+      app.use(baseUrl, express.static(path.resolve(siteDir, 'static')));
+      // TODO: add plugins beforeDevServer hook
```
Comment (endiliey) [inline packages/docusaurus/lib/commands/start.js]: The plugin hooks is something like this in v1

https://github.com/facebook/Docusaurus/blob/94a4ef55c0bfed997b9c7c4d7e32f98695e3cc30/packages/docusaurus-1.x/lib/server/server.js#L347-L349

So for blog plugin, the equivalent is something like

```js
app.use(baseUrl, express.static(path.resolve(siteDir, 'blog', 'assets'));
```

**Comment by yangshun** [comment]
This looks more like it! Finally we have hot reload :)
